server/src/main/java/org/elasticsearch/index/IndexSettingProvider.java

```diff
@@ -22,9 +23,17 @@
      *
      * @param indexName         The name of the new index being created
      * @param dataStreamName    The name of the data stream if the index being created is part of a data stream otherwise <code>null</code>
+     * @param matchingTemplate  The composable index template that matches with the data stream name.
      * @param metadata          The current metadata instance that doesn't yet contain the index to be created
      * @param resolvedAt        The time the request to create this new index was accepted.
      * @param allSettings       All the setting resolved from the template that matches and any setting defined on the create index request
      */
-    Settings getAdditionalIndexSettings(String indexName, String dataStreamName, Metadata metadata, long resolvedAt, Settings allSettings);
+    Settings getAdditionalIndexSettings(
+        String indexName,
+        String dataStreamName,
+        ComposableIndexTemplate matchingTemplate,
```

Comment:
I wonder about this, do we need access to the entire template? Maybe we should pass `DataStreamTemplate` into this instead, which limits a little bit more the information that an index setting provider is expected to retrieve. What do you think?

Reply:
👍 makes sense. I think we should limit the information implementations need to as much as possible. I think just supplying IndexMode is sufficient here too (as `templateIndexMode` parameter).